pynav/batch/problem.py

```diff
@@ -0,0 +1,501 @@
+        nu = 2
+        prev_cost = cost
+
+        if self.verbose:
+            header = "Initial cost: " + str(cost)
+            print(header)
+
+        # Main LM loop
+        while (iter_idx < self.max_iters) and (dx > self.step_tol):
+            A_solve = A + mu * sparse.identity(A.shape[0])
+            delta_x = sparse.linalg.spsolve(A_solve, -b).reshape((-1, 1))
+
+            variables_test = {k: v.copy() for k, v in self.variables.items()}
+
+            # Update the variables
+            self._correct_states(delta_x, variables_test)
+
+            # Compute the new value of the cost function after the update
+            e, H, cost = self.compute_error_jac_cost(variables=variables_test)
+
+            gain_ratio = (prev_cost - cost) / (
+                0.5 * delta_x.T @ (mu * delta_x - b)
+            )
+            gain_ratio = gain_ratio.item(0)
+
+            # If the gain ratio is above zero, accept the step
+            if gain_ratio > 0:
+                self.variables = variables_test
+                mu = mu * max(1.0 / 3.0, 1.0 - (2.0 * gain_ratio - 1) ** 3)
+                nu = 2
+
+                e, H, cost = self.compute_error_jac_cost()
+                cost_list.append(cost)
+                prev_cost = cost
+
+                H_spr = sparse.csr_matrix(H)
+
+                A = H_spr.T @ H_spr
+                b = H_spr.T @ e
+                status = "Accepted."
+            else:
+                mu = mu * nu
+                nu = 2 * nu
+                status = "Rejected."
+
+            dx = np.linalg.norm(delta_x)
+
+            if self.verbose:
+                self._display_header(iter_idx + 1, cost, dx, status=status)
+
+            iter_idx += 1
+
+        # After convergence, compute final value of the cost function
+        e, H, cost = self.compute_error_jac_cost()
+        cost_list.append(cost)
+
+        self._cost_history = np.array(cost_list).reshape((-1))
+        self._information_matrix = A
+
+        return self.variables
+
+    def compute_error_jac_cost(
+        self, variables: Dict[Hashable, State] = None
+    ) -> Tuple[np.ndarray, np.ndarray, float]:
+        """Computes the full error vector, Jacobian, and cost of the problem.
+
+        Parameters
+        ----------
+        variables : Dict[Hashable, State], optional
+            Variables, by default None. If None, uses the variables stored in
+            the optimizer.
+
+        Returns
+        -------
+        Tuple[np.ndarray, np.ndarray, float]
+            Error vector, Jacobian, and cost.
+        """
+
+        if variables is None:
+            variables = self.variables
+
+        # Compute the size of the problem if needed
+        if self._size_errors is None or self._size_state is None:
+            self._compute_size_of_problem()
+
+        # Initialize the error vector and Jacobian
+        e = np.zeros((self._size_errors,))
+        H = np.zeros((self._size_errors, self._size_state))
```

Comment:
i think a big slowdown could occur here. We are allocating and initializing memory for the full dense matrix onto to then put it in a sparse matrix. We should consider constructing the sparse matrix directly